tool/tsh/tsh.go

```diff
@@ -2018,20 +2018,58 @@ func reissueWithRequests(cf *CLIConf, tc *client.TeleportClient, reqIDs ...strin
 	if err != nil {
 		return trace.Wrap(err)
 	}
+
 	params := client.ReissueParams{
 		AccessRequests: reqIDs,
 		RouteToCluster: cf.SiteName,
 	}
+
 	// if the certificate already had active requests, add them to our inputs parameters.
 	if len(profile.ActiveRequests.AccessRequests) > 0 {
 		params.AccessRequests = append(params.AccessRequests, profile.ActiveRequests.AccessRequests...)
 	}
+
 	if params.RouteToCluster == "" {
 		params.RouteToCluster = profile.Cluster
 	}
-	if err := tc.ReissueUserCerts(cf.Context, params); err != nil {
+
+	// Look up any existing key for the cluster. We will need this to check whether
+	// the granted roles have changed post-approval. For our purposes here, it's OK
+	// for this oldKey not to exist.
+
+	oldKey, err := tc.LocalAgent().GetKey(profile.Cluster, client.WithSSHCerts{})
+	switch {
+	case trace.IsNotFound(err):
+		oldKey = nil
+	case err != nil:
 		return trace.Wrap(err)
+	default:
+		// carry on
 	}
+
+	newKey, err := tc.ReissueUserCerts(cf.Context, params)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	// The re-issued certificate attached to newKey may have a different role
+	// set to the old cert, which may in turn change the user's permissions
+	// with *any* service in the cluster (e.g. Kubernetes, Database, etc).
+	//
+	// Check to see if there is a mismatch between the roles granted by the
+	// old key (if any) and the new  key, then delete the user certificates
+	// for the cluster if such a mismatch is detected.
+	//
+	// NB: The SSH cert will have been explicitly updated by the re-issue
+	//     process, so we don't want to delete that. All of the rest are
+	//     fair game.
+
+	err = deleteUserCertsOnRoleMismatch(oldKey, newKey, newKey.ClusterName,
+		tc.LocalAgent(), client.WithAllCertsExceptSSH...)
```

Comment:
As @andrejtokarcik mentioned, we have multiple SSH certs now, one for each cluster.
`newKey` will contain the new SSH cert.
How about deleting _all_ certs, including SSH ones, and then writing `newKey` again?
I believe `tc.ActivateKey(cf.Context, newKey)` should do that

Reply:
I'll give it a shot and let you know.

Reply:
Yep, that's the new approach now. Seems to work in latest patch.